### PR TITLE
better call routing for errors

### PR DESCRIPTION
**Pull request by topepo**
Focuses on functions called by user-facing functions. 

A few more message conversions enabled us to remove glue from depends. 

Also removed an unused function for bartMachine (which we don't support)

### CODE REVIEW

**Comment by topepo** [comment]
After this one, I'll have a corresponding (but trivial) extratests PR ☝️ .

File: R/condense_control.R [hunk 1]
```diff
@@ -20,16 +24,17 @@
...
-condense_control <- function(x, ref) {
+condense_control <- function(x, ref, call = rlang::caller_env()) {
   mismatch <- setdiff(names(ref), names(x))
   if (length(mismatch)) {
     cli::cli_abort(
       c(
-        "Object of class {.cls class(x)[1]} cannot be coerced to
-         object of class {.cls class(ref)[1]}.",
+        "Object of class {.cls {class(x)[1]}} cannot be coerced to
+         object of class {.cls {class(ref)[1]}}.",
```
Comment (simonpcouch) [inline R/condense_control.R:33]: Nice catch!

File: R/condense_control.R [hunk 2]
```diff
@@ -20,16 +24,17 @@
 #'
 #' ctrl <- condense_control(ctrl, control_parsnip())
 #' str(ctrl)
-condense_control <- function(x, ref) {
+condense_control <- function(x, ref, call = rlang::caller_env()) {
```
Comment (simonpcouch) [inline R/condense_control.R:27]:
```suggestion
condense_control <- function(x, ref, ..., call = rlang::caller_env()) {
  check_dots_empty()
```

My reflex here would be to prevent folks from passing `call` by position:

File: R/contr_one_hot.R [hunk 3]
```diff
@@ -24,9 +25,14 @@ contr_one_hot <- function(n, contrasts = TRUE, sparse = FALSE) {
   }
 
   if (is.character(n)) {
+    check_character(n, empty = FALSE)
```
Comment (simonpcouch) [inline R/contr_one_hot.R:28]:
```suggestion
```

I _believe_, in combination with `is.character()` and `length(n) < 1`, remaining cases are covered?

File: R/contr_one_hot.R [hunk 4]
```diff
@@ -24,9 +25,14 @@ contr_one_hot <- function(n, contrasts = TRUE, sparse = FALSE) {
...
+    if (length(n) < 1) {
+      cli::cli_abort("A character vector for {.arg n} cannot be empty.")
```
Comment (simonpcouch) [inline R/contr_one_hot.R:30]:
```suggestion
      cli::cli_abort("{.arg n} cannot be empty.")
```

via https://style.tidyverse.org/errors.html#problem-statement

File: R/arguments.R [hunk 5]
```diff
@@ -258,7 +258,7 @@ make_form_call <- function(object, env = NULL) {
 }
 
 # TODO we need something to indicate that case weights are being used.
-make_xy_call <- function(object, target, env) {
+make_xy_call <- function(object, target, env, call = rlang::caller_env()) {
```
Comment (simonpcouch) [inline R/arguments.R:261]: I'll just make this comment once to apply throughout the PR, but rlang is imported with `@import rlang` so we don't need to namespace rlang for these!

File: R/glmnet-engines.R [hunk 6]
```diff
@@ -379,27 +381,32 @@ format_glmnet_multinom_class <- function(pred, penalty, lvl, n_obs) {
...
-      cli::cli_abort(c(
-        "{.arg penalty} should be a single numeric value.",
-        "i" = "{.fn multi_predict} can be used to get multiple predictions per row of data."
-      ))
+      cli::cli_abort(
+        c(
+          "{.arg penalty} should be a single numeric value.",
+          "i" = "{.fn multi_predict} can be used to get multiple predictions per row of data."
+        ),
+        call = call)
```
Comment (simonpcouch) [inline R/glmnet-engines.R:398]:
```suggestion
        call = call
      )
```

File: R/glmnet-engines.R [hunk 7]
```diff
@@ -379,27 +381,32 @@ format_glmnet_multinom_class <- function(pred, penalty, lvl, n_obs) {
...
-    cli::cli_abort(c(
-      "The glmnet model was fit with a single penalty value of
+    cli::cli_abort(
+      c(
+        "The glmnet model was fit with a single penalty value of
       {.arg object$fit$lambda}. Predicting with a value of {.arg penalty}
       will give incorrect results from `glmnet()`."
-    ))
+      ),
+      call = call)
```
Comment (simonpcouch) [inline R/glmnet-engines.R:409]:
```suggestion
      call = call
    )
```

File: R/predict_class.R [hunk 8]
```diff
@@ -10,13 +10,15 @@
 #' @export
 predict_class.model_fit <- function(object, new_data, ...) {
   if (object$spec$mode != "classification") {
-    cli::cli_abort("{.fun predict.model_fit} is for predicting factor outcomes.")
+    cli::cli_abort("{.fun predict.model_fit} is for predicting factor outcomes.",
+                   call = rlang::call2("predict"))
```
Comment (simonpcouch) [inline R/predict_class.R:14]: This function is `@keywords internal`--I'm game to pass `call` as an argument if you want!

Comment (simonpcouch) [reply R/predict_class.R:14]: Similar comment for other `predict_*.model_fit()` functions, but seems like there may be a rationale there.

File: R/rand_forest_data.R [hunk 9]
```diff
@@ -66,7 +66,8 @@ ranger_confint <- function(object, new_data, ...) {
     } else {
       cli::cli_abort(
         "Cannot compute confidence intervals for a ranger forest
-         of type {.val {object$fit$forest$treetype}}."
+         of type {.val {object$fit$forest$treetype}}.",
+        call = rlang::call2("predict")
```
Comment (simonpcouch) [inline R/rand_forest_data.R:70]: Given that this is an internal helper, I would expect a `call = caller_env()` added to the signature of this one?

File: R/translate.R [hunk 10]
```diff
@@ -176,6 +176,7 @@ add_methods <- function(x, engine) {
 #' or in a list that can facilitate renaming grid objects?
 #' @return A tibble with columns `user`, `parsnip`, and `engine`, or a list
 #' with named character vectors `user_to_parsnip` and `parsnip_to_engine`.
+#' @keywords internal
```
Comment (simonpcouch) [inline R/translate.R:179]: This seems unrelated?

File: tests/testthat/_snaps/sparsevctrs.md [hunk 11]
```diff
@@ -3,7 +3,7 @@
     Code
       xgb_fit <- fit(spec, avg_price_per_room ~ ., data = hotel_data)
     Condition
-      Error in `.convert_form_to_xy_fit()`:
+      Error in `fit()`:
```
Comment (simonpcouch) [inline tests/testthat/_snaps/sparsevctrs.md:6]: looooooooooove it

File: tests/testthat/test-condense_control.R [hunk 12]
```diff
@@ -18,4 +18,12 @@ test_that("condense_control works", {
...
     condense_control(control_parsnip(), ctrl)
   )
+
+  # Emulate being called from one of the upstream control_* functions
+  control_test <- function(control = control_parsnip()) {
+    control <- parsnip::condense_control(control_parsnip(), control)
+    invisible(control)
+  }
+  expect_snapshot(error = TRUE, control_test(ctrl))
+
```
Comment (simonpcouch) [inline tests/testthat/test-condense_control.R:28]:
```suggestion
```

File: tests/testthat/test-contr_one_hot.R [hunk 13]
```diff
@@ -0,0 +1,21 @@
+test_that('one-hot encoding contrasts', {
+
```
Comment (simonpcouch) [inline tests/testthat/test-contr_one_hot.R:2]:
```suggestion
```

File: tests/testthat/test-contr_one_hot.R [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  expect_equal(rownames(chr_contr_mat), letters[1:12])
+  expect_true(all(apply(chr_contr_mat, 1, sum) == 1))
+  expect_true(all(apply(chr_contr_mat, 2, sum) == 1))
+
+  expect_snapshot(contr_one_hot(character(0)), error = TRUE)
+  expect_snapshot(contr_one_hot(-1), error = TRUE)
+  expect_snapshot(contr_one_hot(list()), error = TRUE)
+  expect_snapshot(contr_one_hot(2, contrast = FALSE))
+  expect_snapshot(contr_one_hot(2, sparse = TRUE))
+
```
Comment (simonpcouch) [inline tests/testthat/test-contr_one_hot.R:20]:
```suggestion
```

File: tests/testthat/test-misc.R [hunk 15]
```diff
@@ -5,6 +5,8 @@ test_that('parsnip objects', {
 
   lm_idea <- linear_reg() %>% set_engine("lm")
   expect_false(has_multi_predict(lm_idea))
+  expect_snapshot(error = TRUE, predict(lm_idea, mtcars))
+
```
Comment (simonpcouch) [inline tests/testthat/test-misc.R:9]:
```suggestion
```

File: R/glmnet-engines.R [hunk 16]
```diff
@@ -357,18 +357,20 @@ format_glmnet_multinom_class <- function(pred, penalty, lvl, n_obs) {
...
+    cli::cli_abort(
+      c(
+        "x" = "For the glmnet engine, {.arg penalty} must be a single number
+        (or a value of {.fn tune}).",
+        "!" = "There are {length(pen)} value{?s} for {.arg penalty}.",
+        "i" = "To try multiple values for total regularization, use the
+        {.pkg tune} package.",
+        "i" = "To predict multiple penalties, use {.fn multi_predict}."
+      ),
+      call = call)
```
Comment (simonpcouch) [inline R/glmnet-engines.R:373]:
```suggestion
      call = call
    )
```

**Review by simonpcouch — APPROVED** [review]
HECK YEAH

File: R/arguments.R (hunk 5 above)
Comment (topepo) [reply R/arguments.R:261]: It all comes back to psock... we started namespacing everything in these packages because importing was not enough.

File: R/condense_control.R (hunk 1 above)
Comment (topepo) [reply R/condense_control.R]: Even better, I changed it to `{.obj_type_friendly {x}} `
